filters_test.go

```diff
@@ -346,6 +348,68 @@ func TestMatch(t *testing.T) {
 	}
 }
 
+func TestReadInput(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		Description string
+		Input       string
+		Default     string
+		Want        string
+		ShouldFail  bool
+	}{
+		{
+			Description: "User Input OK",
+			Input:       "123",
+			Default:     "Unused",
+			Want:        "123",
+		},
+		{
+			Description: "No User Input",
+			Input:       "",
+			Default:     "default",
+			Want:        "default",
+		},
+		{
+			Description: "Complex Input",
+			Input:       "/user/test @@@ testing one two three",
+			Default:     "default",
+			Want:        "/user/test @@@ testing one two three",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.Description, func(t *testing.T) {
+			// Prepare Stdin before calling prompt.
+			tmp, err := ioutil.TempFile("", tc.Description)
+			if err != nil {
+				t.Error(err)
+			}
+			defer os.Remove(tmp.Name())
+			if _, err := tmp.WriteString(tc.Input); err != nil {
+				t.Error(err)
+			}
+
+			if _, err := tmp.Seek(0, 0); err != nil {
+				t.Error(err)
+			}
+
+			// Restore original Stdin.
+			oldStdin := os.Stdin
```

Comment:
This definitely isn't concurrency-safe! Have a look at `TestStdin`, for example, for another approach. As I say, I'd like to solve the whole stdin/stdout issue globally for pipes, but we'll get there.

Reply:
I am all for solving the problem in a different way. However, If we decide against doing the solution discussed earlier (the one with two versions). I'm all for stubbing `os.Stdin` and using a `mutex` to ensure concurrency-safety. 

The suggested solution in `TestStdin` while apparently concurrency-safe seemed a bit impure (dependency on external shell commands if I understood correctly). While the task is admittedly complex, I am really all for solving it differently. 

I ran `TestReadInput` with Golang's race detector with the command below, and got no errors. I am no expert in concurrency but I found it strange that the race detector didn't detect the error, or is it just that it wasn't made to discover that exact type.
```bash
go test -count=1 -race -v -run TestReadInput 
```

Reply:
The race detector can only detect races that actually _occur_ while running the tests, not potential races. I think if you create _another_ parallel test that tries to write to `os.Stdin`, for example, you might trigger it.

Okay, it sounds like this is where we are: we don't want to do stdin-related testing in a crap way, and we don't want to put a bunch of engineering effort into doing it a better way that won't be needed once we've solved the overall issue of pipe-level stdin/stdout. So we should probably solve that _first_, and then come back to this PR and use it.

Are you up for having a go?

Reply:
Great points there. I guess the latest commit addresses these issues.